File: src/core/suppressions.ts
import { EnergyViolation, VIOLATION_TYPE } from '../types';

// decision: parsed by scanning raw source lines with a regex instead of walking the tree-sitter
// AST — the comment node type differs per grammar (see LanguageAdapter.nodeTypes.comment) but the
// esa-ignore marker text itself is identical across languages, so a text scan avoids needing a
// per-language comment-node lookup just to find `//` and `#` comments.
// decision: anchored with `\s*$` so the marker must be a complete token — bare at end-of-line or
// followed by `:` + type list — not prose that merely mentions "esa-ignore" (a file whose own
// comment starts "// esa-ignore ..." would otherwise parse as an unused bare directive).
const DIRECTIVE_PATTERN = /(\/\/|#)\s*esa-ignore(-file)?(?::\s*([\w,\s-]+))?\s*$/;

const KNOWN_TYPES: Set<string> = new Set(Object.values(VIOLATION_TYPE));

export interface Suppression {
    // 0-indexed line the directive comment itself sits on (matches EnergyViolation.line).
    line: number;
    column: number;
    scope: 'line' | 'file';
    types: string[] | 'all';
    unknownTypes: string[];
    // Whether the directive is the only content on its line — see isCoveredBy below.
    standalone: boolean;
}

// decision: an explicit but entirely-unrecognized type list (e.g. a typo'd `esa-ignore:
// nseting`) resolves to an empty list, not a fallback to 'all' — falling back to 'all' would
// make a typo silently suppress *more* than intended instead of less, which is the opposite of
// the "keep suppression honest" goal (see the unused-directive note this produces below). Only
// a bare directive with no colon at all means "suppress every type".
function parseTypeList(raw: string | undefined): { types: string[] | 'all'; unknownTypes: string[] } {
    if (!raw) {
        return { types: 'all', unknownTypes: [] };
    }
    const tokens = raw
        .split(',')
        .map((token) => token.trim())
        .filter(Boolean);
    const types = tokens.filter((token) => KNOWN_TYPES.has(token));
    const unknownTypes = tokens.filter((token) => !KNOWN_TYPES.has(token));
    return { types, unknownTypes };
}

export function parseSuppressions(sourceText: string): Suppression[] {
    const lines = sourceText.split('\n');
    const suppressions: Suppression[] = [];

    lines.forEach((lineText, line) => {
        const match = DIRECTIVE_PATTERN.exec(lineText);
        if (!match) {
            return;
        }
        const { types, unknownTypes } = parseTypeList(match[3]);
        suppressions.push({
            line,
            column: match.index,
            scope: match[2] ? 'file' : 'line',
            types,
            unknownTypes,
            standalone: lineText.slice(0, match.index).trim() === ''
        });
    });

    return suppressions;
}

// decision: a standalone directive (nothing but the comment on its line) also covers the next
// line — lets a suppression read naturally placed above a multi-line construct's header (a
// function signature, an if-condition) instead of forcing it onto an already-long code line.
function coversLine(suppression: Suppression, violationLine: number): boolean {
    if (suppression.scope === 'file') {
        return true;
    }
    if (violationLine === suppression.line) {
        return true;
    }
    return suppression.standalone && violationLine === suppression.line + 1;
}

function matchesType(suppression: Suppression, type: string): boolean {
    return suppression.types === 'all' || suppression.types.includes(type);
}

export interface ApplySuppressionsResult {
    violations: EnergyViolation[];
    suppressionNotes: EnergyViolation[];
}

// decision: returns unused/unknown-type directives as their own low-severity violations (rather
// than silently dropping them) — an esa-ignore that stopped matching anything (the violation was
// fixed, or the type name was mistyped) is exactly the kind of stale suppression that should be
// visible, not a silent no-op that quietly keeps working "by accident".
export function applySuppressions(violations: EnergyViolation[], sourceText: string): ApplySuppressionsResult {
    const suppressions = parseSuppressions(sourceText);
    if (suppressions.length === 0) {
        return { violations, suppressionNotes: [] };
    }

    const suppressedCounts = new Map<Suppression, number>(suppressions.map((s) => [s, 0]));

    const remaining = violations.filter((violation) => {
        const match = suppressions.find((s) => coversLine(s, violation.line) && matchesType(s, violation.type));
        if (!match) {
            return true;
        }
        suppressedCounts.set(match, (suppressedCounts.get(match) ?? 0) + 1);
        return false;
    });

    const suppressionNotes: EnergyViolation[] = [];
    for (const suppression of suppressions) {
        if (suppression.unknownTypes.length > 0) {
            suppressionNotes.push({
                line: suppression.line,
                column: suppression.column,
                type: VIOLATION_TYPE.SUPPRESSION,
                severity: 'low',
                message: `esa-ignore names unknown violation type(s): ${suppression.unknownTypes.join(', ')}.`
            });
        }
        if ((suppressedCounts.get(suppression) ?? 0) === 0) {
            const scopeText = suppression.scope === 'file' ? 'file-wide ' : '';
            suppressionNotes.push({
                line: suppression.line,
                column: suppression.column,
                type: VIOLATION_TYPE.SUPPRESSION,
                severity: 'low',
                message: `Unused ${scopeText}esa-ignore — no matching violation found. Remove it or fix the type list.`
            });
        }
    }

    return { violations: remaining, suppressionNotes };
}

File: src/Core/Suppressions.fs
module Energy.Core.Suppressions

open System.Text.RegularExpressions
open Energy.Core.Violation

// decision: these are the capture-group indices of directivePattern (group 1 = comment marker,
// group 2 = optional `-file`, group 3 = suppression types); named constants make the indexed reads
// below self-documenting instead of bare magic numbers. They sit at the top of the module so this
// choice is visible rather than hidden next to their use sites.
let private scopeGroupIndex = 2
let private typeGroupIndex = 3

type SuppressionScope =
    | Line
    | File

type Suppression =
    { Line: int
      Column: int
      Scope: SuppressionScope
      Types: Set<ViolationType> option
      UnknownTypes: string list
      Standalone: bool }

type ApplySuppressionsResult =
    { Violations: EnergyViolation list
      SuppressionNotes: EnergyViolation list }

let private directivePattern =
    Regex("(//|#)\\s*esa-ignore(-file)?(?::\\s*([\\w,\\s-]+))?\\s*$")

// decision: directive type names mirror `violationTypeName` (the JSON/report contract) rather than
// the detector's internal `Name`, so a user copies the exact string from any report. Multi-word
// types are kebab-case (`primitive-obsession`), matching docs and CLI output.
let private knownTypes =
    [ "nesting", Nesting
      "complexity", Complexity
      "cognitive", Cognitive
      "naming", Naming
      "coherence", Coherence
      "magic", Magic
      "parameters", Parameters
      "inversion", Inversion
      "primitive-obsession", PrimitiveObsession
      "match-opportunity", MatchOpportunity
      "logical-control-flow", LogicalControlFlow
      "opaque-boolean", OpaqueBoolean
      "suppression", Suppression ]
    |> Map.ofList

// Parses a type list conservatively: a bare directive means every type, but a list containing only
// misspellings means no type at all and therefore cannot accidentally suppress unrelated findings.
let private parseTypeList (raw: string) =
    if System.String.IsNullOrWhiteSpace raw then
        None, []
    else
        let tokens =
            raw.Split(',')
            |> Array.map _.Trim()
            |> Array.filter (fun token -> token <> "")
            |> Array.toList

        let types, unknownTypes =
            tokens
            |> List.fold
                (fun (types, unknownTypes) token ->
                    match Map.tryFind token knownTypes with
                    | Some violationType -> Set.add violationType types, unknownTypes
                    | None -> types, unknownTypes @ [ token ])
                (Set.empty, [])

        Some types, unknownTypes

// decision: scans source text rather than AST comment nodes because the directive marker is identical
// across languages while each tree-sitter grammar represents comments differently.
let parseSuppressions (sourceText: string) : Suppression list =
    sourceText.Split('\n')
    |> Array.mapi (fun line lineText ->
        let matched = directivePattern.Match lineText

        if not matched.Success then
            None
        else
            let types, unknownTypes = parseTypeList matched.Groups.[typeGroupIndex].Value

            Some
                { Line = line
                  Column = matched.Index
                  Scope =
                    if matched.Groups.[scopeGroupIndex].Success then
                        File
                    else
                        Line
                  Types = types
                  UnknownTypes = unknownTypes
                  Standalone = lineText.Substring(0, matched.Index).Trim() = "" })
    |> Array.choose id
    |> Array.toList

let private coversLine suppression violationLine =
    match suppression.Scope with
    | File -> true
    | Line ->
        violationLine = suppression.Line
        || (suppression.Standalone && violationLine = suppression.Line + 1)

let private matchesType suppression violationType =
    suppression.Types |> Option.forall (Set.contains violationType)

// decision: emits a low-severity finding for unknown or unused directives so suppression debt is
// visible; filtering a violation is never allowed to turn a stale comment into a silent no-op.
let applySuppressions (violations: EnergyViolation list) (sourceText: string) : ApplySuppressionsResult =
    let suppressions = parseSuppressions sourceText |> List.indexed

    if suppressions.IsEmpty then
        { Violations = violations
          SuppressionNotes = [] }
    else
        let remaining, suppressedCounts =
            violations
            |> List.fold
                (fun (remaining, counts) violation ->
                    match
                        suppressions
                        |> List.tryFind (fun (_, suppression) ->
                            coversLine suppression violation.Line && matchesType suppression violation.Type)
                    with
                    | Some(index, _) -> remaining, Map.change index (Option.defaultValue 0 >> (+) 1 >> Some) counts
                    | None -> violation :: remaining, counts)
                ([], Map.empty)

        let notes =
            suppressions
            |> List.collect (fun (index, suppression) ->
                let unknownNote =
                    if suppression.UnknownTypes.IsEmpty then
                        []
                    else
                        [ { Line = suppression.Line
                            Column = suppression.Column
                            Type = Suppression
                            Severity = Low
                            Message =
                              sprintf
                                  "esa-ignore names unknown violation type(s): %s."
                                  (System.String.Join(", ", suppression.UnknownTypes))
                            Hotspots = [] } ]

                let unusedNote =
                    if Map.tryFind index suppressedCounts |> Option.defaultValue 0 > 0 then
                        []
                    else
                        let scopeText = if suppression.Scope = File then "file-wide " else ""

                        [ { Line = suppression.Line
                            Column = suppression.Column
                            Type = Suppression
                            Severity = Low
                            Message =
                              sprintf
                                  "Unused %sesa-ignore — no matching violation found. Remove it or fix the type list."
                                  scopeText
                            Hotspots = [] } ]

                unknownNote @ unusedNote)

        { Violations = List.rev remaining
          SuppressionNotes = notes }
